File: src/modules/EventBus.ts
/* eslint-disable @typescript-eslint/no-unsafe-argument */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-empty-function */
/* eslint-disable @typescript-eslint/no-explicit-any */


import { MessageId, Messages } from "./modules_const";
import * as event from 'event.event';

/*
    шина сообщений
*/

declare global {
    const EventBus: ReturnType<typeof EventBusModule>;
}


export function register_event_bus() {
    if (!event)
        return Log.error('Не подключена библиотека event.event');
    event.set_logger({ error: Log.error });
    (_G as any).EventBus = EventBusModule();
}

type FncOnCallback<T> = (data: T) => void;

interface ListenerInfo {
    callback: FncOnCallback<any>;
    callback_context: any;
    url: hash;
    once: boolean;
    id_script: string;
    is_message_mode: boolean;
}

function EventBusModule() {
    const bus_log = Log.get_with_prefix('Bus');
    const events: { [key_message: string]: IEventClass } = {};
    const listeners: { [key_message: number]: ListenerInfo[] } = {};
    const send_messages: { [k: number]: { time: number, data: any, listener: ListenerInfo } } = {};
    let id_counter_message_bus = 0;
    const EVENT_BUS_MESSAGE = hash('event_bus_message');

    function ensure_hash(string_or_hash: string | hash) {
        return (type(string_or_hash) == "string" ? hash(string_or_hash as string) : string_or_hash) as number; // real is hash
    }

    function get_id_script() {
        const url: any = msg.url();
        return hash_to_hex(url.socket) + hash_to_hex(url.path) + hash_to_hex(url.fragment || hash(""));
    }

    function update_cache() {
        for (const k in send_messages) {
            const message = send_messages[k];
            if (message.time + 2 < System.now())
                delete send_messages[k];
        }
    }

    function _on<T extends MessageId>(id_message: T, callback: FncOnCallback<Messages[T]>, is_message_mode: boolean, callback_context: any, once: boolean) {
        //bus_log.log('on', id_message, is_message_mode);
        const key_message = ensure_hash(id_message);
        if (is_message_mode) {
            //
        }
        else {
            if (!events[key_message])
                events[key_message] = event.create();
            events[key_message].subscribe(callback);
        }
        if (!listeners[key_message])
            listeners[key_message] = [];
        listeners[key_message].push({ callback, callback_context, once, is_message_mode, url: msg.url(), id_script: get_id_script() });
    }

    function on<T extends MessageId>(id_message: T, callback: FncOnCallback<Messages[T]>, is_message_mode = false, callback_context?: any) {
        _on(id_message, callback, is_message_mode, callback_context, false);
    }

    function once<T extends MessageId>(id_message: T, callback: FncOnCallback<Messages[T]>, is_message_mode = false, callback_context?: any) {
        _on(id_message, callback, is_message_mode, callback_context, true);
    }

    function off<T extends MessageId>(id_message: T, callback: FncOnCallback<Messages[T]>) {
        const key_message = ensure_hash(id_message);
        if (!listeners[key_message]) {
            bus_log.warn(`Ни один слушатель для события не зарегистрирован: ${id_message}, off`);
            return;
        }
        const list = listeners[key_message];
        for (let i = list.length - 1; i >= 0; i--) {
            const l = list[i];
            if (l.callback == callback) {
                if (!l.is_message_mode && events[key_message])
                    events[key_message].unsubscribe(callback);
                list.splice(i, 1);
                return;
            }
        }
    }

    function off_all_id_message<T extends MessageId>(id_message: T) {
        const key_message = ensure_hash(id_message);
        if (!listeners[key_message]) {
            bus_log.warn(`Ни один слушатель для события не зарегистрирован: ${id_message}, off_all_id_message`);
            return;
        }
        if (events[key_message]) {
            events[key_message].clear();
            delete events[key_message];
        }
        delete listeners[key_message];
    }

    function off_all_context(context: any) {
        for (const key_message in listeners) {
            const listener = listeners[key_message];
            for (let i = listener.length - 1; i >= 0; i--) {
                const l = listener[i];
                if (l.callback_context == context) {
                    if (!l.is_message_mode && events[key_message])
                        events[key_message].unsubscribe(l.callback);
                    listener.splice(i, 1);
                }
            }
        }
    }

    function off_all_current_script() {
        const id_script = get_id_script();
        for (const key_message in listeners) {
            const listener = listeners[key_message];
            for (let i = listener.length - 1; i >= 0; i--) {
                const l = listener[i];
                if (l.id_script == id_script) {
                    if (!l.is_message_mode && events[key_message])
                        events[key_message].unsubscribe(l.callback);
                    listener.splice(i, 1);
                }
            }
        }
    }

    function trigger<T extends MessageId>(id_message: T, message_data?: Messages[T], show_warning = true, is_copy_data = false) {
        const key_message = ensure_hash(id_message);
        if (!listeners[key_message]) {
            if (show_warning)
                bus_log.warn(`Ни один слушатель для события не зарегистрирован: ${id_message}, trigger/send`);
            return;
        }
        const data = is_copy_data ? json.decode(json.encode(message_data)) : message_data; // чтобы во всех случаях была копия(редко когда нужно иначе)
        // важный момент для случая once, что сначала происходит тригер, а затем удаление события, т.е. вешать событие внутри колбека нужно аккуратно учитывая это
        const list = listeners[key_message];
        const del_ids = [];
        for (let i = 0; i < list.length; i++) {
            const l = list[i];
            if (l.is_message_mode) {
                id_counter_message_bus++;
                send_messages[id_counter_message_bus] = { time: System.now(), data, listener: l };
                msg.post(l.url, EVENT_BUS_MESSAGE, { id_counter_message_bus });
            }
            if (l.once)
                del_ids.push(i);
        }
        if (events[key_message])
            events[key_message].trigger(data);

        for (let i = del_ids.length - 1; i >= 0; i--) {
            const id = del_ids[i];
            const l = list[id];
            if (!l.is_message_mode && events[key_message])
                events[key_message].unsubscribe(l.callback);
            list.splice(id, 1);
        }
    }

    function send<T extends MessageId>(id_message: T, message_data?: Messages[T]) {
        return trigger(id_message, message_data);
    }

    function on_message(_this: any, id_message: hash, _message: any, sender: hash) {
        if (id_message == EVENT_BUS_MESSAGE) {
            const id_script = get_id_script();
            const message = send_messages[_message.id_counter_message_bus];
            const listener = message.listener;
            if (listener.id_script == id_script)
                listener.callback(message.data);
            update_cache();
        }
    }

    return { on, once, off, off_all_id_message, off_all_context, off_all_current_script, on_message, send, trigger };
}

File: app/modules/EventBus.lua
local ____lualib = require("lualib_bundle")
local __TS__Delete = ____lualib.__TS__Delete
local __TS__ArraySplice = ____lualib.__TS__ArraySplice
local ____exports = {}
local EventBusModule
local event = require("event.event")
function EventBusModule()
    local bus_log = Log.get_with_prefix("Bus")
    local events = {}
    local listeners = {}
    local send_messages = {}
    local id_counter_message_bus = 0
    local EVENT_BUS_MESSAGE = hash("event_bus_message")
    local function ensure_hash(string_or_hash)
        return type(string_or_hash) == "string" and hash(string_or_hash) or string_or_hash
    end
    local function get_id_script()
        local url = msg.url()
        return (hash_to_hex(url.socket) .. hash_to_hex(url.path)) .. hash_to_hex(url.fragment or hash(""))
    end
    local function update_cache()
        for k in pairs(send_messages) do
            local message = send_messages[k]
            if message.time + 2 < System.now() then
                __TS__Delete(send_messages, k)
            end
        end
    end
    local function _on(id_message, callback, is_message_mode, callback_context, once)
        local key_message = ensure_hash(id_message)
        if is_message_mode then
        else
            if not events[key_message] then
                events[key_message] = event.create()
            end
            events[key_message]:subscribe(callback)
        end
        if not listeners[key_message] then
            listeners[key_message] = {}
        end
        local ____listeners_key_message_0 = listeners[key_message]
        ____listeners_key_message_0[#____listeners_key_message_0 + 1] = {
            callback = callback,
            callback_context = callback_context,
            once = once,
            is_message_mode = is_message_mode,
            url = msg.url(),
            id_script = get_id_script()
        }
    end
    local function on(id_message, callback, is_message_mode, callback_context)
        if is_message_mode == nil then
            is_message_mode = false
        end
        _on(
            id_message,
            callback,
            is_message_mode,
            callback_context,
            false
        )
    end
    local function once(id_message, callback, is_message_mode, callback_context)
        if is_message_mode == nil then
            is_message_mode = false
        end
        _on(
            id_message,
            callback,
            is_message_mode,
            callback_context,
            true
        )
    end
    local function off(id_message, callback)
        local key_message = ensure_hash(id_message)
        if not listeners[key_message] then
            bus_log.warn(("Ни один слушатель для события не зарегистрирован: " .. id_message) .. ", off")
            return
        end
        local list = listeners[key_message]
        do
            local i = #list - 1
            while i >= 0 do
                local l = list[i + 1]
                if l.callback == callback then
                    if not l.is_message_mode and events[key_message] then
                        events[key_message]:unsubscribe(callback)
                    end
                    __TS__ArraySplice(list, i, 1)
                    return
                end
                i = i - 1
            end
        end
    end
    local function off_all_id_message(id_message)
        local key_message = ensure_hash(id_message)
        if not listeners[key_message] then
            bus_log.warn(("Ни один слушатель для события не зарегистрирован: " .. id_message) .. ", off_all_id_message")
            return
        end
        if events[key_message] then
            events[key_message]:clear()
            __TS__Delete(events, key_message)
        end
        __TS__Delete(listeners, key_message)
    end
    local function off_all_context(context)
        for key_message in pairs(listeners) do
            local listener = listeners[key_message]
            do
                local i = #listener - 1
                while i >= 0 do
                    local l = listener[i + 1]
                    if l.callback_context == context then
                        if not l.is_message_mode and events[key_message] then
                            events[key_message]:unsubscribe(l.callback)
                        end
                        __TS__ArraySplice(listener, i, 1)
                    end
                    i = i - 1
                end
            end
        end
    end
    local function off_all_current_script()
        local id_script = get_id_script()
        for key_message in pairs(listeners) do
            local listener = listeners[key_message]
            do
                local i = #listener - 1
                while i >= 0 do
                    local l = listener[i + 1]
                    if l.id_script == id_script then
                        if not l.is_message_mode and events[key_message] then
                            events[key_message]:unsubscribe(l.callback)
                        end
                        __TS__ArraySplice(listener, i, 1)
                    end
                    i = i - 1
                end
            end
        end
    end
    local function trigger(id_message, message_data, show_warning, is_copy_data)
        if show_warning == nil then
            show_warning = true
        end
        if is_copy_data == nil then
            is_copy_data = false
        end
        local key_message = ensure_hash(id_message)
        if not listeners[key_message] then
            if show_warning then
                bus_log.warn(("Ни один слушатель для события не зарегистрирован: " .. id_message) .. ", trigger/send")
            end
            return
        end
        local ____is_copy_data_1
        if is_copy_data then
            ____is_copy_data_1 = json.decode(json.encode(message_data))
        else
            ____is_copy_data_1 = message_data
        end
        local data = ____is_copy_data_1
        local list = listeners[key_message]
        local del_ids = {}
        do
            local i = 0
            while i < #list do
                local l = list[i + 1]
                if l.is_message_mode then
                    id_counter_message_bus = id_counter_message_bus + 1
                    send_messages[id_counter_message_bus] = {
                        time = System.now(),
                        data = data,
                        listener = l
                    }
                    msg.post(l.url, EVENT_BUS_MESSAGE, {id_counter_message_bus = id_counter_message_bus})
                end
                if l.once then
                    del_ids[#del_ids + 1] = i
                end
                i = i + 1
            end
        end
        if events[key_message] then
            events[key_message]:trigger(data)
        end
        do
            local i = #del_ids - 1
            while i >= 0 do
                local id = del_ids[i + 1]
                local l = list[id + 1]
                if not l.is_message_mode and events[key_message] then
                    events[key_message]:unsubscribe(l.callback)
                end
                __TS__ArraySplice(list, id, 1)
                i = i - 1
            end
        end
    end
    local function send(id_message, message_data)
        return trigger(id_message, message_data)
    end
    local function on_message(_this, id_message, _message, sender)
        if id_message == EVENT_BUS_MESSAGE then
            local id_script = get_id_script()
            local message = send_messages[_message.id_counter_message_bus]
            local listener = message.listener
            if listener.id_script == id_script then
                listener.callback(message.data)
            end
            update_cache()
        end
    end
    return {
        on = on,
        once = once,
        off = off,
        off_all_id_message = off_all_id_message,
        off_all_context = off_all_context,
        off_all_current_script = off_all_current_script,
        on_message = on_message,
        send = send,
        trigger = trigger
    }
end
function ____exports.register_event_bus()
    if not event then
        return Log.error("Не подключена библиотека event.event")
    end
    event.set_logger({error = Log.error})
    _G.EventBus = EventBusModule()
end
return ____exports
